データベース版メモアプリ

yama998877 opened the pull request:
pgを使用してPostgreSQLに接続できるように変更。
PostgreSQLとの接続を使いまわせるように変更。
メモデータの保存に使用していたファイル操作のコードをデータベース操作のコードに変更。
プレースホルダを用いてクエリを作成し、SQLインジェクションをブロックするように変更。
メモファイル操作時の変数名を修正。
readmeにデータベース、テーブルの作成方法追加。gemのインストール方法の修正。
erbファイルの変数名を修正。

yama998877 committed:
データベース版メモアプリ

pgを使用してPostgreSQLに接続できるように変更。
接続を使いまわせるように変更。
メモデータの保存に使用していたファイル操作のコードをデータベース操作のコードに修正。
プレースホルダを用いてクエリを作成し、SQLインジェクションをブロックするように変更。
メモファイル操作時の変数名を修正。
readmeにデータベース、テーブルの作成方法追加。gemのインストール方法の修正。
erbファイルの変数名を修正。

yoshitsugu commented on .gitignore:
```diff
@@ -0,0 +1 @@
+vendor
```
これは何のためのgitignoreでしょうか。

yama998877 replied:
インストールしたruby gemの保存先ディレクトリを除外するためのgitignoreです。
変更のあったファイルとして大量に表示されたり、うっかり変更したファイルをすべてpushするのを防ぐために追加しました🙇‍♂️

yoshitsugu replied:
デフォルトの設定で `rbenv` などを使っていれば不要(このディレクトリ以下のvendorに保存されることはない)かと思うのですが、このディレクトリにrubygemsを保存する必要がありますか？
Rubyのプロジェクトを追加する度に毎回gitignoreするのも大変かと思うので、なしで済ませられるならなしでよいかと思います。

yama998877 replied:
ご指摘ありがとうございます。gemの保存先を変更していたようです。申し訳ありません。デフォルトに戻してリポジトリ内のディレクトリ内に保存されないようにしました。gitignoreファイルは不要になったので削除しました。

yoshitsugu commented on app.rb:
```diff
@@ -2,38 +2,47 @@
-    memo = JSON.parse(File.read(f), symbolize_names: true)
-    uuid = File.basename(f, '.json')
-    memos[uuid] = memo
+def read_memos(memo_id = 'id,title,detail')
```
`id` に `id` でないものを無理矢理渡さないようにしましょう。
もしこういうコードがライブラリとしてあったときに利用者としてはどういう使い方をしたらいいのか、コードをちゃんと読んで挙動を把握しないといけないです。メソッド名や引数などのインターフェースを見てある程度推測できるようなものにすると親切です。
Rubyは自由度が高いので、こういう恣意的なコードを書けるのですが、書き捨てのものならともかく、実務で使うコードではこういう書き方をするとメンテナンス性がとても下がるので今のうちから避けられるようにしましょう。

yama998877 replied:
ご指摘ありがとうございます🙇‍♂️
`read_memos(memo_id ='none')`に変更して引数がなかった場合の処理だとわかりやすいように修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -2,38 +2,47 @@
+def read_memos(memo_id = 'id,title,detail')
+  if memo_id == 'id,title,detail'
+    memos = {}
+    CONN.exec("SELECT #{memo_id} FROM memos") do |result|
```
ここの `#{memo_id} ` でSQLインジェクションの恐れがあります。そもそもここを動的にする意味があまりないと思うのですが、どうでしょうか。

yama998877 replied:
ご指摘ありがとうございます🙇‍♂️
デフォルト値の`memo_id = 'id,title,detail'`を使用した方がいいと思って動的になりました。
`if memo_id == 'id,title,detail'`を`read_memos(memo_id ='none')`に修正したので`#{memo_id}`を修正しました。

yoshitsugu reviewed (commented):
コメントしました。

yama998877 committed:
指摘のあった箇所の修正

`.gitignorea`ファイルに何のディレクトリを除外するのか説明コメントを追加。
app.rbのread_memosメソッドのデフォルト値、クエリの修正

yoshitsugu commented on app.rb:
```diff
@@ -2,38 +2,47 @@
+require 'pg'
 set :enviroment, :production
 
+CONN = PG.connect(dbname: 'memoapp')
```
DBのコネクションは途中でクローズされたり状態が変わり得ますので、あまり定数で管理するのは適切ではありません。変数でよいのではないでしょうか。

yama998877 replied:
ご指摘ありがとうございます。変数に修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -2,38 +2,47 @@
-    memo = JSON.parse(File.read(f), symbolize_names: true)
-    uuid = File.basename(f, '.json')
-    memos[uuid] = memo
+def read_memos(memo_id = 'none')
+  if memo_id == 'none'
+    memos = {}
+    CONN.exec('SELECT id,title,detail FROM memos') do |result|
+      result.each do |row|
+        memo = { title: row['title'], detail: row['detail'] }
+        uuid = row['id']
+        memos[uuid] = memo
+      end
+    end
+  else
+    memos = 0
+    CONN.exec_params('SELECT title,detail FROM memos WHERE id = $1', [memo_id]) do |result|
+      result.each do |row|
+        memo = { title: row['title'], detail: row['detail'] }
+        memos = memo
+      end
+    end
```
`memo_id` が `none` かどうかによって挙動がかわるのもかなりトリッキーです。
そもそも `id` を指定したい場合と全体を取得したい場合はやりたいことが違うと思うので、メソッドを分けてはどうでしょうか。
無理に一つのメソッドでやる必要はないと思います。
```ruby
# 全体取得用
def read_memos
end

# id指定での取得用
def read_memo(memo_id)
end
```

yama998877 replied:
ご指摘ありがとうございます。無理に一つに収めようとして申し訳ありません。メソッドを分けました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -2,38 +2,47 @@
+    end
+  else
+    memos = 0
+    CONN.exec_params('SELECT title,detail FROM memos WHERE id = $1', [memo_id]) do |result|
```
一件だけを取得したい場合はSQLレベルで `LIMIT 1` をつけるようにして余計な件数を取得できないようにしましょう。

yama998877 replied:
ご指摘ありがとうございます。`LIMIT 1`をクエリに追加しました🙇‍♂️

yoshitsugu reviewed (commented):
コメントしました。

yama998877 committed:
修正

定数CONNでのデータベース接続を変数connに変更。
read_memosメソッドをread_memos,read_memo(memo_id)に切り分け。
read_memo(memo_id)メソッド内のクエリにLIMIT 1を追加。
修正したメソッドに対応できるようにコードを修正。

yama998877 committed:
変数でのデータベース接続でメソッドが動かなくなったので修正。

各メソッドの引数にconnを追加。
メモの一覧表示ができなくなったのでインスタンス変数`@all_memo`に変更。

yama998877 committed:
gitignoreの削除

gemの保存先をデフォルトに戻したので削除

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
-  File.open(filename, 'w') do |file|
-    file.puts(memo.to_json)
+def read_memo(conn, memo_id)
+  memo = 0
```
ここで `= 0` しているのはなぜでしょうか。
`Hash` または `0` がはいる、というのはあまり直感的ではないです。たとえば該当のメモがない場合は `nil` を返す、ではどうでしょうか。

yama998877 replied:
ご指摘ありがとうございます。`nil`があることを忘れていてとりあえず値を入れなければと思い、0を入れました。申し訳ありません。修正しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
+end
+
+def write_memo(conn, uuid, memo_title, memo_detail)
+  if read_memos(conn).key?(uuid) == false
```
`== false` は `!(条件)` を使いましょう。
https://docs.ruby-lang.org/ja/latest/doc/spec=2foperator.html#not

yama998877 replied:
ご指摘ありがとうございます。`!= ture`に修正しました🙇‍♂️

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
+end
+
+def write_memo(conn, uuid, memo_title, memo_detail)
+  if read_memos(conn).key?(uuid) == false
+    conn.exec_params('INSERT INTO memos VALUES ($1,$2,$3,now())', [uuid, memo_title, memo_detail])
+  else
+    conn.exec('UPDATE memos SET (title, detail, update_at) = ($1,$2, now()) WHERE id = $3', [memo_title, memo_detail, uuid])
   end
```
まとめてもよいのですが、今回の場合、 `POST` と `PUT` で入口自体が分かれているので、あまりここをまとめるメリットもないかなと思いました。
`create_memo` と `update_memo` に分割してよいと思います。

yama998877 replied:
ご指摘ありがとうございます。修正してメソッドを分割しました。

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
 end
 
 get '/memos' do
+  @all_memo = read_memos(conn)
```
変数名は `@memos` くらいでよいと思います。基本、Rubyでは複数形でArrayをあらわすことが多いです。

yama998877 replied:
ご指摘ありがとうございます。修正しました。

yoshitsugu reviewed (commented):
コメントしました！

yama998877 committed:
修正

read_memoメソッド内の変数`memo`の値を0からnilに変更。
write_memoメソッドをcreate_memoメソッドとupdate_memoメソッドに分割。
write_memoメソッドを使用していた箇所をcreate_memo,update_memoに変更。
@all_memoを@memosに変更。

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
+end
+
+def create_memo(conn, uuid, memo_title, memo_detail)
+  conn.exec_params('INSERT INTO memos VALUES ($1,$2,$3,now())', [uuid, memo_title, memo_detail]) if read_memos(conn).key?(uuid) != true
```
`if !read_memos(conn).key?(uuid)` もしくは `unless read_memos(conn).key?(uuid)` でよさそうです。
`if` の条件として `true` や `false` と `==` 、 `!=` してしまっている場合、それは不要ではないか、検討してみてください。

yama998877 replied:
ご指摘ありがとうございます。前回のコメントの
> `== false` は `!(条件)` を使いましょう。

の意図に気づかず、`!='に修正してしまい申し訳ありませんでした🙇 

`unless read_memos(conn).key?(uuid)`に修正しました。

yoshitsugu commented on app.rb:
```diff
@@ -2,47 +2,57 @@
-    memos[uuid] = memo
+  conn.exec('SELECT id,title,detail FROM memos') do |result|
+    result.each do |row|
+      memo = { title: row['title'], detail: row['detail'] }
+      uuid = row['id']
+      memos[uuid] = memo
```
```suggestion
      memos[row['id']] = { title: row['title'], detail: row['detail'] }
```

yama998877 replied:
ご指摘ありがとうございます。
`memos[row['id']] = { title: row['title'], detail: row['detail'] }`に修正して、23行〜25行は削除しました。

yoshitsugu reviewed (commented):
コメントしました！

yama998877 committed:
メソッド内のコード修正,require 'pathname'の削除

read_memosメソッド内のコードを修正。
create_memoメソッド内の条件式をunlessを使うように修正。

yoshitsugu reviewed (approved):
OKです！
